highlight-range.js

```diff
@@ -1,19 +1,48 @@
 var highlightRange = (function () {
-// Wrap each text node in a given DOM Range with a <span class=[highLightClass]>.
+// Wrap each text node in a given DOM Range with an inline node , defaulting to <span class=[highLightClass]>.
 // Breaks start and/or end node if needed.
 // Returns a function that cleans up the created highlight (not a perfect undo: split text nodes are not merged again).
 //
 // Parameters:
 // - rangeObject: a Range whose start and end containers are text nodes.
-// - highlightClass: the CSS class the text pieces in the range should get, defaults to 'highlighted-range'.
-function highlightRange(rangeObject, highlightClass) {
+// - highlightTemplate
+//     string: the CSS class the text pieces in the range should get, defaults to 'highlighted-range'
+//     HTMLElement: an element that will be cloned for each highlight
+//     function: a function that creates a wrapper node. It's passed the Node that's being
+//               highligted (but is free to ignore it). This can be used to apply properties 
+//               to a logical highlight, even if it's spread over many elements, for example
+//               to store an annotation identifier to be able to query for all highlight
+//               spans for the same annotation:
+//               
+//                 function renderAnnotation(annotation, range) {
+//                   highlightRange(range, function() {
+//                     var span = document.createElement('span');
+//                     span.classList.add('annotation');
+//                     span.dataset.annotationId = annotation.id;
+//                     return span;
+//                   });
+//                 }
+function highlightRange(rangeObject, highlightTemplate) {
     // Ignore range if empty.
     if (rangeObject.collapsed) {
         return;
     }
 
-    if (typeof highlightClass == 'undefined') {
-        highlightClass = 'highlighted-range';
+    var highlightCallback = highlightTemplate || defaultHighlight;
```

Comment:
I would remove this line, and add extra `if`s, getting something like this (possibly in different order):
```
if (typeof highlightTemplate === 'function') {
    highlightCallback = highlightTemplate
}
if (typeof highlightTemplate === 'undefined') {
  highlightCallback = node => defaultHighlight(node, 'highlighted-range')
}
```
(last line assumes that my other suggestion, to change defaultHighlight's signature, is applied)